File: optimus-fe/src/types/assets.ts
// Shapes mirror the backend assets DTOs. Field names intentionally remain
// snake_case so the TypeScript contracts match the JSON wire format directly.

export interface CloudAccountSummary {
  id: number
  name: string
  provider: 'aws'
  cloudkey_id: number
  cloudkey_name: string
  regions_count: number
  enabled: boolean
  last_sync_at?: string
  last_sync_status?: string
  created_at: string
  updated_at: string
}

export interface CloudAccountDetail extends CloudAccountSummary {
  enabled_regions: string[]
  description: string
}

export interface CreateCloudAccountRequest {
  name: string
  provider: 'aws'
  cloudkey_id: number
  enabled_regions: string[]
  enabled?: boolean
  description?: string
}

export interface UpdateCloudAccountRequest {
  name?: string
  enabled_regions?: string[]
  enabled?: boolean
  description?: string
}

export interface InstanceSummary {
  id: number
  cloud_account_id: number
  cloud_account_name?: string
  region: string
  instance_id: string
  name: string
  instance_type: string
  state: string
  private_ip?: string
  public_ip?: string
  vpc_id?: string
  subnet_id?: string
  availability_zone?: string
  launch_time?: string
  tags: Record<string, string>
  last_seen_at: string
  deleted: boolean
}

export interface VPCSummary {
  id: number
  cloud_account_id: number
  cloud_account_name?: string
  region: string
  vpc_id: string
  name: string
  cidr_block?: string
  is_default: boolean
  state: string
  tags: Record<string, string>
  last_seen_at: string
  deleted: boolean
}

export interface SubnetSummary {
  id: number
  cloud_account_id: number
  region: string
  subnet_id: string
  vpc_id: string
  name: string
  cidr_block?: string
  availability_zone: string
  tags: Record<string, string>
  last_seen_at: string
  deleted: boolean
}

export interface DatabaseSummary {
  id: number
  cloud_account_id: number
  cloud_account_name?: string
  region: string
  db_instance_id: string
  engine: string
  engine_version: string
  instance_class: string
  status: string
  endpoint: string
  port?: number
  multi_az: boolean
  publicly_accessible: boolean
  storage_gb?: number
  tags: Record<string, string>
  last_seen_at: string
  deleted: boolean
}

export type SyncRunStatus = 'running' | 'success' | 'failed' | 'skipped'
export type SyncRunResourceType = 'instance' | 'network' | 'database'
export type SyncRunTrigger = 'cron' | 'manual' | 'test'

export interface SyncRunSummary {
  id: number
  cloud_account_id: number
  cloud_account_name?: string
  region: string
  resource_type: SyncRunResourceType
  started_at: string
  finished_at?: string
  status: SyncRunStatus
  items_seen: number
  items_softdeleted: number
  error?: string
  error_code?: number
  trigger: SyncRunTrigger
  triggered_by_user_id?: number
}

export interface ListResponse<T> {
  items: T[]
  total: number
}

File: optimus-be/internal/models/assets.go
package models

import (
	"net"
	"time"

	"gorm.io/datatypes"
	"gorm.io/gorm"
)

type CloudAccount struct {
	ID             uint64      `gorm:"primaryKey"`
	Name           string      `gorm:"size:128;not null"`
	Provider       string      `gorm:"size:16;not null"`
	CloudKeyID     uint64      `gorm:"column:cloudkey_id;not null;index"`
	EnabledRegions StringArray `gorm:"type:text[];not null"`
	Enabled        bool        `gorm:"not null;default:true"`
	Description    string      `gorm:"type:text;not null;default:''"`
	CreatedAt      time.Time
	UpdatedAt      time.Time
	DeletedAt      gorm.DeletedAt `gorm:"index"`
}

func (CloudAccount) TableName() string { return "cloud_accounts" }

type AWSInstance struct {
	ID               uint64  `gorm:"primaryKey"`
	CloudAccountID   uint64  `gorm:"column:cloud_account_id;not null"`
	Region           string  `gorm:"size:32;not null"`
	InstanceID       string  `gorm:"column:instance_id;size:32;not null"`
	Name             string  `gorm:"type:text;not null;default:''"`
	InstanceType     string  `gorm:"size:32;not null;default:''"`
	State            string  `gorm:"size:16;not null;default:''"`
	PrivateIP        *net.IP `gorm:"type:inet"`
	PublicIP         *net.IP `gorm:"type:inet"`
	VPCID            string  `gorm:"column:vpc_id;size:32;not null;default:''"`
	SubnetID         string  `gorm:"column:subnet_id;size:32;not null;default:''"`
	AvailabilityZone string  `gorm:"size:32;not null;default:''"`
	LaunchTime       *time.Time
	Tags             datatypes.JSON `gorm:"type:jsonb;not null;default:'{}'"`
	LastSeenAt       time.Time      `gorm:"not null"`
	CreatedAt        time.Time
	UpdatedAt        time.Time
	DeletedAt        gorm.DeletedAt `gorm:"index"`
}

func (AWSInstance) TableName() string { return "aws_instances" }

type AWSVPC struct {
	ID             uint64         `gorm:"primaryKey"`
	CloudAccountID uint64         `gorm:"column:cloud_account_id;not null"`
	Region         string         `gorm:"size:32;not null"`
	VPCID          string         `gorm:"column:vpc_id;size:32;not null"`
	Name           string         `gorm:"type:text;not null;default:''"`
	CIDRBlock      *string        `gorm:"column:cidr_block;type:cidr"`
	IsDefault      bool           `gorm:"not null;default:false"`
	State          string         `gorm:"size:16;not null;default:''"`
	Tags           datatypes.JSON `gorm:"type:jsonb;not null;default:'{}'"`
	LastSeenAt     time.Time      `gorm:"not null"`
	CreatedAt      time.Time
	UpdatedAt      time.Time
	DeletedAt      gorm.DeletedAt `gorm:"index"`
}

func (AWSVPC) TableName() string { return "aws_vpcs" }

type AWSSubnet struct {
	ID               uint64         `gorm:"primaryKey"`
	CloudAccountID   uint64         `gorm:"column:cloud_account_id;not null"`
	Region           string         `gorm:"size:32;not null"`
	SubnetID         string         `gorm:"column:subnet_id;size:32;not null"`
	VPCID            string         `gorm:"column:vpc_id;size:32;not null"`
	CIDRBlock        *string        `gorm:"column:cidr_block;type:cidr"`
	AvailabilityZone string         `gorm:"size:32;not null;default:''"`
	Name             string         `gorm:"type:text;not null;default:''"`
	Tags             datatypes.JSON `gorm:"type:jsonb;not null;default:'{}'"`
	LastSeenAt       time.Time      `gorm:"not null"`
	CreatedAt        time.Time
	UpdatedAt        time.Time
	DeletedAt        gorm.DeletedAt `gorm:"index"`
}

func (AWSSubnet) TableName() string { return "aws_subnets" }

type AWSDatabase struct {
	ID                 uint64 `gorm:"primaryKey"`
	CloudAccountID     uint64 `gorm:"column:cloud_account_id;not null"`
	Region             string `gorm:"size:32;not null"`
	DBInstanceID       string `gorm:"column:db_instance_id;size:64;not null"`
	Engine             string `gorm:"size:32;not null;default:''"`
	EngineVersion      string `gorm:"size:32;not null;default:''"`
	InstanceClass      string `gorm:"size:32;not null;default:''"`
	Status             string `gorm:"size:32;not null;default:''"`
	Endpoint           string `gorm:"type:text;not null;default:''"`
	Port               *int32
	MultiAZ            bool           `gorm:"column:multi_az;not null;default:false"`
	PubliclyAccessible bool           `gorm:"not null;default:false"`
	StorageGB          *int32         `gorm:"column:storage_gb"`
	Tags               datatypes.JSON `gorm:"type:jsonb;not null;default:'{}'"`
	LastSeenAt         time.Time      `gorm:"not null"`
	CreatedAt          time.Time
	UpdatedAt          time.Time
	DeletedAt          gorm.DeletedAt `gorm:"index"`
}

func (AWSDatabase) TableName() string { return "aws_databases" }

type AssetsSyncRun struct {
	ID                uint64 `gorm:"primaryKey"`
	CloudAccountID    uint64 `gorm:"column:cloud_account_id;not null"`
	Region            string `gorm:"size:32;not null"`
	ResourceType      string `gorm:"size:16;not null"`
	StartedAt         time.Time
	FinishedAt        *time.Time
	Status            string `gorm:"size:16;not null"`
	ItemsSeen         int32  `gorm:"not null;default:0"`
	ItemsSoftDeleted  int32  `gorm:"column:items_softdeleted;not null;default:0"`
	Error             string `gorm:"type:text;not null;default:''"`
	ErrorCode         int32  `gorm:"not null;default:0"`
	Trigger           string `gorm:"size:16;not null"`
	TriggeredByUserID *uint64
	CreatedAt         time.Time
}

func (AssetsSyncRun) TableName() string { return "assets_sync_runs" }
